Payment driver rotessa

karneaud opened the pull request:
Integration of Rotessa Payment Gateway

karneaud committed:
update providers with Rotess Provider

karneaud committed:
add migration to add rotessa gateway

karneaud committed:
add package

karneaud committed:
update aliases

karneaud committed:
add appropriate class

karneaud committed:
update gateway types

karneaud committed:
refactor classes to facilitiate import customers

karneaud committed:
set default country instead

karneaud committed:
fix the data passed

karneaud committed:
fix placholder text

karneaud committed:
refactor methods to properly get accounts and set tokens

karneaud committed:
commit package

karneaud committed:
Merge pull request #1 from karneaud/v5-develop

V5 develop

karneaud committed:
Merge pull request #2 from karneaud/v5-develop

V5 develop

karneaud committed:
remove

karneaud committed:
add data providers for privinces and installment

karneaud committed:
add default attributes for rotessa gateway views

karneaud committed:
hanlde attributes for rotessa views via components

karneaud committed:
add rotessa gateway payment methods

karneaud committed:
update

karneaud committed:
add service provider for rotessa gateway

karneaud committed:
add rotessa service provider to config

karneaud committed:
add migration to add rotessa payment gateway

karneaud committed:
add views for rotessa payment gateway

karneaud committed:
add rotessa dependency class maps

karneaud committed:
fix class name

karneaud committed:
remove composer package

karneaud committed:
add rotessa sdk

karneaud committed:
allow matching of any client contact

karneaud committed:
Merge pull request #3 from karneaud/v5-develop

V5 develop

karneaud committed:
move componetn class to composer provider

karneaud committed:
remove rotessa service provider

karneaud committed:
reference full blade path using render

karneaud committed:
use alias

karneaud committed:
Merge branch 'payment-driver-rotessa' into v5-develop

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Merge pull request #4 from karneaud/v5-develop

V5 develop

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/invoiceninja/invoiceninja?pullRequest=9704) <br/>All committers have signed the CLA.

karneaud committed:
Merge branch 'invoiceninja:v5-develop' into v5-develop

karneaud committed:
do not need these files

karneaud committed:
removed this file

karneaud committed:
refactor compser components for rotessa

karneaud committed:
Merge pull request #6 from invoiceninja/v5-develop

V5 develop

karneaud committed:
Merge pull request #7 from invoiceninja/v5-develop

V5 develop

karneaud committed:
Merge pull request #5 from invoiceninja/v5-develop

V5 develop

karneaud committed:
Merge branch 'payment-driver-rotessa' into v5-develop

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Merge pull request #8 from karneaud/v5-develop

V5 develop

karneaud committed:
update import process to import/ sync new/existing customers

karneaud commented:
Updated branch to include revisions for importing all customers and syncing existing customers

karneaud committed:
Merge branch 'invoiceninja:v5-develop' into v5-develop

turbo124 commented on resources/views/portal/ninja2020/gateways/rotessa/bank_transfer/CA/details.blade.php:
```diff
@@ -0,0 +1,4 @@
+<dd> Gateway: </dd>
+<dt>{{ $brand }}</dt>
+<dd> Account Number: </dd>
```
add translation references

karneaud replied:
Code is outdated. This has been updated

karneaud committed:
Merge pull request #15 from M-E-Development-Design/v5-develop

V5 develop

karneaud committed:
Update texts.php

Updated french canadian language texts

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update account.blade.php

updated to handle translations

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update texts.php

update canadian language text

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update address.blade.php

Updated to use language translation features

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update texts.php

update text translation for canada

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update contact.blade.php

update blade template to use language translation features

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update US.blade.php

Update blade to use the language translation feature

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update CA.blade.php

update blade to use the language translation feature

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update bank.blade.php

update blade to use language translation feature

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update bank.blade.php

Update blade to use language translation feature

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update details.blade.php

Update blade to use language translation feature

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud commented:
@turbo124  Updated translation texts for fr_CA and ca. Updated blades to use language translation features

turbo124 commented on app/Models/Gateway.php:
```diff
@@ -105,7 +105,9 @@ public function getHelp()
         } elseif ($this->id == 62) {
             $link = 'https://docs.btcpayserver.org/InvoiceNinja/';
-        }
+        } elseif ($this->id == 4002) {
```
To ensure we increment accurately, we'll assign the ID 63 to this gateway, if you can adjust accordingly please.

karneaud replied:
Amended as per request

turbo124 commented on app/PaymentDrivers/Rotessa/PaymentMethod.php:
```diff
@@ -0,0 +1,215 @@
+            $customer = ClientGatewayToken::query()
+                ->where('company_gateway_id', $this->rotessa->company_gateway->id)
+                ->where('client_id', $this->rotessa->client->id)
+                ->where('id', (int) $request->input('token_id'))
```
we use the obfuscated id on the front end rather than the integer ID. do you would need to "decode" the token ID at this point (and ensure that you are not using the int in the front end.

```php
            ->where('id', $this->decodePrimaryKey($request->token))
```

karneaud replied:
Removed the use of the id entirely

turbo124 commented on app/PaymentDrivers/Rotessa/PaymentMethod.php:
```diff
@@ -0,0 +1,215 @@
+            [ 'data' => $data ],
+            SystemLog::CATEGORY_GATEWAY_RESPONSE,
+            SystemLog::EVENT_GATEWAY_SUCCESS,
+            880,
```
For readability, you may want to add ROTESSA as a const in SystemLog

karneaud replied:
Added constant as per request

turbo124 commented on app/PaymentDrivers/Rotessa/PaymentMethod.php:
```diff
@@ -0,0 +1,215 @@
+            $this->rotessa->client->company,
+        );
+
+        return redirect()->route('client.payments.show', [ 'payment' => $this->rotessa->encodePrimaryKey($payment->id) ]);
```
this can be replaced with

['payment' => $payment->hashed_id]

karneaud replied:
Updated as per request

turbo124 commented on app/PaymentDrivers/RotessaPaymentDriver.php:
```diff
@@ -0,0 +1,276 @@
+            $client_contacts = $customers->filter(function ($value, $key) use ($client_emails) {
+                return !in_array(((object) $value)->email, $client_emails);
+            })->each( function($customer) use ($company_id) {
+                sleep(10);
```
can the sleep() be removed?

karneaud replied:
Removed as per request

turbo124 commented on database/migrations/2024_06_11_231143_add_rotessa_gateway.php:
```diff
@@ -0,0 +1,56 @@
+    public function up(): void
+    {
+        Model::unguard();
+        \DB::statement('SET FOREIGN_KEY_CHECKS=0;');
```
no need for the foreign key checks here.

also, you may want to use 

database/migrations/2023_06_20_123355_add_paypal_rest_payment_driver.php

as a template, the gateway->key field needs to be a static string, the one in this migration will create a new static string per installation. we do use the ->key elsewhere in the application, so it needs to be uniform.

you will also need to add the gateway array to the PaymentLibrarySeeders.php

karneaud replied:
Refactored as per request and added to payment library seeder

turbo124 commented on database/migrations/2024_06_11_231143_add_rotessa_gateway.php:
```diff
@@ -0,0 +1,56 @@
+        $gateway->default_gateway_type_id = 2;
+        $gateway->save();
+
+ 	    Gateway::query()->where('name','=', 'Rotessa')->update(['visible' => 1]);
```
safe to remove this.

karneaud replied:
Removed as per request

turbo124 commented on database/migrations/2024_06_11_231143_add_rotessa_gateway.php:
```diff
@@ -0,0 +1,56 @@
+     */
+    public function down(): void
+    {
+         Gateway::where('name', '=', 'Rotessa')->delete();
```
safe to remove this.

karneaud replied:
Done as per request

turbo124 commented on lang/ca/texts.php:
```diff
@@ -5301,6 +5301,23 @@
     'latest_requires_php_version' => 'Note: the latest version requires PHP :version',
     'auto_expand_product_table_notes' => 'Automatically expand products table notes',
     'auto_expand_product_table_notes_help' => 'Automatically expands the notes section within the products table to display more lines.',
+    'institution_number' => 'Institution Number',
+    'transit_number' => 'Transit Number',
+    'personal' => 'Personal',
+    'province_code' => 'Province Code',
+    'province_abbreviation' => 'Province Abbreviation',
+    'street_address' => 'Street Address',
+    'full_name' => 'Full Name',
+    'address_information' => 'Address Information',
+    'enter_the_information_for_the_bank_account' => 'Enter the Information for the Bank Account',
+    'address_line_1' => 'Address Line 1',
+    'address_line_2' => 'Address Line 2',
+    'account_holder_information' => 'Account Holder Information',
+    'enter_information_for_the_account_holder' => 'Enter Information for the Account Holder',
+    'home_phone' => 'Home Phone',
+    'customer_type' => 'Customer Type',
+    'process_date' => 'Process Date',
+    'other_phone' => 'Other Phone'
```
only the en/texts.php file is needed, also there are duplicate here, we need to reuse any existing translations and only add in ones that are not already present in the translation file.

karneaud replied:
Not sure what these duplicates are. Can you point them out to me. Note that I notice this pull request isn't using the most recent code updates. Should I make a new one?

turbo124 replied:
![image](https://github.com/user-attachments/assets/f7ff7c85-4dda-4095-9ecd-0dec913c285c)

We always try to reuse / keep consistent the translations. We already have ~5k translation so we try and be as prudent as possible

karneaud replied:
Hmmm. I see. The text was based on Rotessa's fields and the fact that it uses "Canadian English Language" This may differ in context from US/ other english language. If you are suggesting I use the "other language" as a base instead I would but would this not affect the context and language translations?

turbo124 replied:
We only support English and English - United Kingdom. 

The translations you have added here have been put in the CA language which is Catalan. Also for any language other than en, we manage the translations in Transifex here https://transifex.com

If English - Canadian as a language wants to be added, this would need to be added into Transifex, and then the translations would need to be perfomed on all strings until they were at least 70% completed prior to importing that in as a new language.

karneaud replied:
Ok I see. So to be clear...

-  I need to update either **en_US** or **en_GB** language file.
-  I should also try to use what's already there? e.g **'texts.address1'** as oppose to **'texts.address_line_1'**
-  I should try to use a _close match_ as oppose to adding a **custom match** where possible? e.g. "postal_code' => 'Postcode' instead of me adding **'ca_postal_code' => 'Postal Code'** ?

Would need to report back so just need to be sure.

turbo124 replied:
@karneaud only the en/texts.php file is ever updates with new translations, thats correct. The remaining points are also correct. If you find a match in the texts.php file, please use it, if there is no close match, then the new translation can be added to en/texts.php

turbo124 commented on lang/fr_CA/texts.php:
```diff
@@ -5299,6 +5299,24 @@
     'auto_expand_product_table_notes' => 'Développer automatiquement les notes du tableau de produits',
     'auto_expand_product_table_notes_help' => ' 
 Développe automatiquement la section des notes dans le tableau de produits pour afficher plus de lignes.',
+    'transit_number'=>'Numéro de transit',
+    'institution_number' => 'Numéro d\'institution',
+    'account_type' => 'Type de compte',
+    'personal' => 'Personnel',
+    'province_code' => 'Code de la province (abréviation)',
+    'province_abbreviation' => 'Abréviation de la province',
+    'postal_code' => 'Code postal',
+    'street_address' => 'Adresse',
+    'full_name' => 'Nom complet',
+    'address_information' => 'Informations de l\'adresse',
+    'enter_the_information_for_the_bank_account' => 'Entrez les informations du compte bancaire',
+    'address_line_1' => 'Adresse ligne 1',
+    'address_line_2' => 'Adresse ligne 2',
+    'account_holder_information' => 'Informations du titulaire du compte',
+    'enter_information_for_the_account_holder' => 'Entrez les informations du titulaire du compte',
+    'home_phone' => 'Téléphone résidentiel',
+    'customer_type' => 'Type de client',
+    'process_date' => 'Date de traitement'
```
remove this

karneaud replied:
I have removed it completely

turbo124 commented on resources/views/portal/ninja2020/gateways/rotessa/bank_transfer/pay.blade.php:
```diff
@@ -0,0 +1,91 @@
+            @if (count($tokens) > 0)
+                @foreach ($tokens as $token)
+                    <label class="mr-4">
+                        <input type="radio" data-token="{{ $token->token }}" data-token_id="{{ $token->id }}" name="payment-type"
```
token->hashed_id

karneaud replied:
Removed it completely

turbo124 commented on resources/views/portal/ninja2020/gateways/rotessa/bank_transfer/pay.blade.php:
```diff
@@ -0,0 +1,91 @@
+            </dd>
+            {{--
+            <dt class="text-sm leading-5 font-medium text-gray-500 mr-4">
+                Insallments
+            </dt>
+            <dd class="mt-1 text-sm leading-5 text-gray-900 sm:mt-0 sm:col-span-2">
+                <input class="input w-full" id="installments" name="installments" type="number" placeholder="Installments" required value="{{ old('installments',$installments) }}">
+            </dd>
+            
+            <dt class="text-sm leading-5 font-medium text-gray-500 mr-4">
+                Frequency
+            </dt>
+            <dd class="mt-1 text-sm leading-5 text-gray-900 sm:mt-0 sm:col-span-2">
+                <input class="input w-full" id="frequency" name="frequency" type="text" placeholder="Once/Weekly/Monthly/Annually" required >
+            </dd>
+            
+            <dt class="text-sm leading-5 font-medium text-gray-500 mr-4">
+                Comments
```
needs to be replaced with ctrans()....

karneaud replied:
Amended as per request

turbo124 reviewed (changes requested):
Thanks, few small changes needed.

karneaud committed:
Update Gateway.php

Increment id

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update pay.blade.php

don't use token id integer on front end

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update PaymentMethod.php

updated to not use token id integer but decrypt it

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update PaymentMethod.php

remove id from sql

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update SystemLog.php

add rotessa to system log type

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update PaymentMethod.php

Update to use systemlog const

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update PaymentMethod.php

use $payment property value

_Originally posted by @turbo124 in https://github.com/invoiceninja/invoiceninja/pull/9704#discussion_r1693677928_


Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update RotessaPaymentDriver.php

removed sleep mode as per request https://github.com/invoiceninja/invoiceninja/pull/9704#discussion_r1693679459

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update 2024_06_11_231143_add_rotessa_gateway.php

Reefactored code as per request https://github.com/invoiceninja/invoiceninja/pull/9704/files/2ea54b0042b247ffdb054f7f707544a2b145ed23#r1693681342

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update PaymentLibrariesSeeder.php

added Rotessa to seeder library as per request https://github.com/invoiceninja/invoiceninja/pull/9704#discussion_r1693681455

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update 2024_06_11_231143_add_rotessa_gateway.php

remove tear down as per request https://github.com/invoiceninja/invoiceninja/pull/9704#discussion_r1693681455

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update texts.php

remove manual update as per request https://github.com/invoiceninja/invoiceninja/pull/9704#discussion_r1693681870

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update pay.blade.php

removed. Not needed

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update pay.blade.php

Added ctrans() features as per request https://github.com/invoiceninja/invoiceninja/pull/9704#discussion_r1693682268

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update Gateway.php

Updated to correct id

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

turbo124 committed:
Add new Gateway

turbo124 committed:
Ensure api credentials are camelCase

turbo124 committed:
Fixes for missing '

karneaud committed:
Update texts.php

Moved to en/texts.php

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

turbo124 commented on app/PaymentDrivers/RotessaPaymentDriver.php:
```diff
@@ -0,0 +1,274 @@
+        return $this;
+    }
+
+    public function gatewayTypes(): array
+    {
+        $types = [];
+
+        if ($this->client
+        && $this->client->currency()
+        && in_array($this->client->currency()->code, ['USD'])
+        && isset($this->client->country)
+        && in_array($this->client->country->iso_3166_2, ['US'])) {
+            $types[] = GatewayType::BANK_TRANSFER;
+        }
+
+        if ($this->client
+            && $this->client->currency()
+            && in_array($this->client->currency()->code, ['CAD'])
+            && isset($this->client->country)
+            && in_array($this->client->country->iso_3166_2, ['CA'])) {
+            $types[] = GatewayType::ACSS;
+        }
+        
+        return $types;
+    }
```
Please ensure that the supported payment methods applicable for this implementation are curated here.

ie. CAD + CAN only payments.

karneaud replied:
Updated based on recommendations

turbo124 reviewed (changes requested):
Small changes

karneaud committed:
Update pay.blade.php

update text translation

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update CA.blade.php

amend translation text ref

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update address.blade.php

Amend translation texts

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update contact.blade.php

Amend translation text.

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update contact.blade.php

update amendments 

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update pay.blade.php

Remove in favor of internal method to generate comment

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update PaymentMethod.php

use default internal description method.

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update texts.php

update text translations

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update RotessaPaymentDriver.php

Remove until proper testing can be done.

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update account.blade.php

Amend text translations

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

turbo124 committed:
Updated resources

karneaud committed:
Merge branch 'invoiceninja:v5-develop' into v5-develop

karneaud committed:
Merge branch 'payment-driver-rotessa' into v5-develop

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Merge pull request #16 from M-E-Development-Design/v5-develop

V5 develop

karneaud committed:
add missing defaults

karneaud committed:
fix missing custom identifier property

karneaud committed:
fix missing custom identifier. fix missing translation

karneaud committed:
fix missing translation

karneaud committed:
use system log code

karneaud committed:
fix log values

turbo124 commented on app/PaymentDrivers/RotessaPaymentDriver.php:
```diff
@@ -0,0 +1,276 @@
+        return true;
+    }
+    
+    public function findOrCreateCustomer(array $data)
+    {
+        $result = null; 
+        try {
+           
+            $existing = ClientGatewayToken::query()
+                ->where('company_gateway_id', $this->company_gateway->id)
+                ->where('client_id', $this->client->id)
+                ->orWhere(function (Builder $query) use ($data) {
+                    $query->where('token', encrypt(join(".", Arr::only($data, 'id','custom_identifier')))  )
```
first check for tokens, then client email. If both fail, then create a client.

turbo124 committed:
Fixes for hosted

turbo124 commented:
Pending some final QA adjustments prior to merge @karneaud

karneaud committed:
Update RotessaPaymentDriver.php

Set payment_method_id to use ACSS. Updates meta data.

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>

karneaud committed:
Update PaymentMethod.php

Updated to use ACSS for CAN clients only.

Signed-off-by: Kendall Arneaud <kendall.arneaud@gmail.com>
